linux kernel: prefer zstd where possible

Ma27 opened the pull request:
## Description of changes

Closes #302291
Closes #301536

The following things have changed:

* For 5.11+: ZRAM compressor uses zstd now.
* For 5.13+: kernel modules are compressed with zstd instead of xz.
* For 5.19+: support zstd-compressed firmware.


<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Ma27 committed:
linux kernel: prefer zstd where possible

Closes #302291
Closes #301536

The following things have changed:

* For 5.7+: ZSWAP compressor uses zstd now.
* For 5.11+: ZRAM compressor uses zstd now.
* For 5.13+: kernel modules are compressed with zstd instead of xz.
* For 5.19+: support zstd-compressed firmware.

The modules-closure functionality needed explicit support for copying
over `.zst` files. Also, the VM image builder used busybox's `insmod`
before which doesn't support zstd. Switched to `kmod` and added xz/zstd
as dependencies for it, similar to how it's done for the actual stage1
in d33e52b2539c5b36a5a876df9006a7145efd42ea. The use of `kmod` here
doesn't seem to be such a big deal since it's only a build-time
dependency.

Ma27 commented:
Going to rebuild my infra on my Hydra and boot into that kernel to see if this has some unexpected side-effects.

Ma27 commented:
First breakage I found is that images from `make-disk-image.nix` are broken, seeing stuff like this in the buildlog:

```
[    0.255498] Invalid ELF header magic: != ELF
[    0.256614] Invalid ELF header magic: != ELF
insmod: can't insert '/nix/store/aq5nbyxd3zsi51z1bahfq2ja07ssivqx-linux-6.1.84-shrunk/lib/module[    0.258685] Invalid ELF header magic: != ELF
s/6.1.84/kernel/[    0.259465] Invalid ELF header magic: != ELF
```

Taking a look.

aviallon commented on pkgs/os-specific/linux/kernel/common-config.nix:
```diff
@@ -868,11 +868,12 @@ let
-      ZBUD           = option yes;
+      ZRAM               = module;
+      ZRAM_WRITEBACK     = option yes;
+      ZRAM_DEF_COMP_ZSTD = whenAtLeast "5.11" yes;
```
```suggestion
      ZRAM_DEF_COMP_ZSTD = whenAtLeast "5.11" yes;
      ZSWAP_COMPRESSOR_DEFAULT_ZSTD = whenAtLeast "5.19" yes;
```
If you select `zstd` as the default compressor for Zram, perhaps you should do the same for Zswap (for the sake of consistency).

Ma27 commented:
> Taking a look.

Have a fix locally now, but I'm too tired now to push, will try to get to it tomorrow.

alois31 commented:
Has the impact on the size been measured for the kernel modules? The "comparable" compression ratio of zstd (as compared to LZMA) can be quite a bit worse (double digit percentage) depending on the levels and specifics of the uncompressed data. (This concern should not apply to the other two points; zram changes from a worse compressor and for the firmware it's only adding support.)

Ma27 commented:
> Has the impact on the size been measured for the kernel modules

fwiw the stage1 modules-shrunk is 7.6M (xz) vs 9.4M (zstd) on my laptop.
the `kernel-modules` path of the same machine is 305M (xz) vs 339M (zstd).

Currently preparing a patch that provides zstd compression of the firmware. Here we have 492.6M (zstd) vs 440.3M (xz) though I'm playing around with different levels here.

Will probably document the exact impact in the commit message and perhaps also in the release notes.

Ma27 committed:
nixos/udev: compress firmware with zstd if possible

Closes #267442

    $ nix path-info -Sh /nix/store/qj1dm7wfw5m3mxf1gn3fdm0az9y1h5ny-linux-firmware-20240312-xz
    /nix/store/qj1dm7wfw5m3mxf1gn3fdm0az9y1h5ny-linux-firmware-20240312-xz	440.3M
    $ nix path-info -Sh /nix/store/c3szcjxb3g990dbiz7llwmkaf0bi98j2-linux-firmware-20240312-zstd
    /nix/store/c3szcjxb3g990dbiz7llwmkaf0bi98j2-linux-firmware-20240312-zstd	460.6M

This is an increase of 4.4%, but OTOH zstd has a significantly higher
decompression speed[1].

[1] https://gregoryszorc.com/blog/2017/03/07/better-compression-with-zstandard/

Ma27 commented:
Pushed an improved version of the original patch that fixes the VM image builder. Pushed another commit which actually uses zstd to compress firmware.

Will rebuild my infra once again with these two patches to see if I find more problems and how well it works out.

There's a ~11% increase (440M -> 490M) in linux firmware now. Perhaps we want to make the behavior configurable here?
I played around a bit with the flags for zstd, but with a very high compression level (15) I only got it down to 483M.

Will write something to the release notes later.

Also cc @fpletz.

alois31 commented:
I'd personally be more worried about the modules, since some of them go into the initrd. If I understand correctly, this is where you observed the increase from 7.6MB to 9.4MB, which seems quite bad.

aviallon commented:
> Pushed an improved version of the original patch that fixes the VM image builder. Pushed another commit which actually uses zstd to compress firmware.
> 
> Will rebuild my infra once again with these two patches to see if I find more problems and how well it works out.
> 
> There's a ~11% increase (440M -> 490M) in linux firmware now. Perhaps we want to make the behavior configurable here? I played around a bit with the flags for zstd, but with a very high compression level (15) I only got it down to 483M.
> 
> Will write something to the release notes later.
> 
> Also cc @fpletz.

You could try the `--long` option, which should work.

aviallon commented on pkgs/build-support/kernel/compress-firmware.nix:
```diff
@@ -1,24 +1,38 @@
+    zstd = {
+      ext = "zst";
+      nativeBuildInputs = [ zstd ];
+      cmd = file: target: ''zstd -T1 -6 --check -f "${file}" -o "${target}"'';
```
```suggestion
      cmd = file: target: ''zstd -T1 -19 --long --check -f "${file}" -o "${target}"'';
```
You can go as high as level 19, and `--long` should work too. This is consistent with what is configured for `xz`.

lheckemann replied:
I wonder if we could reuse/expand the data from `pkgs/build-support/kernel/initrd-compressor-meta.nix`?

fpletz commented:
Personally, I'ld love to see zstd support. But as others mentioned xz generally seems to have a better compression ratio than zstd. I hope the suggestions from @aviallon can make a difference.

The hit on the firmware is IMHO pretty bad since most people running on hardware will need at least one binary blog from the standard linux nonfree firmware package.

I was thinking about splitting it because some of the bigger blobs from unpopular vendors are used very rarely anyway:

```
[…]
   1.86 MiB result/lib/firmware/libertas
   2.15 MiB result/lib/firmware/ueagle-atm
   2.32 MiB result/lib/firmware/iwlwifi-8000C-36.ucode
   2.32 MiB result/lib/firmware/iwlwifi-8265-36.ucode
   2.33 MiB result/lib/firmware/iwlwifi-8265-34.ucode
   2.34 MiB result/lib/firmware/iwlwifi-8000C-34.ucode
   2.36 MiB result/lib/firmware/qca
   2.41 MiB result/lib/firmware/iwlwifi-9000-pu-b0-jf-b0-38.ucode
   2.41 MiB result/lib/firmware/iwlwifi-9260-th-b0-jf-b0-38.ucode
   2.55 MiB result/lib/firmware/iwlwifi-9000-pu-b0-jf-b0-34.ucode
   2.55 MiB result/lib/firmware/iwlwifi-9260-th-b0-jf-b0-34.ucode
   2.58 MiB result/lib/firmware/bnx2x
   3.94 MiB result/lib/firmware/radeon
   4.57 MiB result/lib/firmware/cxgb4
   4.80 MiB result/lib/firmware/liquidio
   6.08 MiB result/lib/firmware/cypress
   6.20 MiB result/lib/firmware/cirrus
   6.52 MiB result/lib/firmware/ath12k
   7.58 MiB result/lib/firmware/rtw89
   8.67 MiB result/lib/firmware/ti-connectivity
  11.99 MiB result/lib/firmware/brcm
  14.74 MiB result/lib/firmware/dpaa2
  19.44 MiB result/lib/firmware/ath10k
  20.43 MiB result/lib/firmware/qed
  25.87 MiB result/lib/firmware/i915
  34.94 MiB result/lib/firmware/mediatek
  43.37 MiB result/lib/firmware/ath11k
  58.95 MiB result/lib/firmware/intel
  60.94 MiB result/lib/firmware/amdgpu
  64.19 MiB result/lib/firmware/nvidia
  80.99 MiB result/lib/firmware/mrvl
  81.63 MiB result/lib/firmware/mellanox
 140.34 MiB result/lib/firmware/netronome
 151.34 MiB result/lib/firmware/qcom
   1.05 GiB total
```

For instance, we could split it loosely based on how Debian is currently doing it: https://packages.debian.org/search?suite=sid&searchon=names&keywords=firmware

Ma27 commented:
> I was thinking about splitting it because some of the bigger blobs from unpopular vendors are used very rarely anyway:

fwiw that works pretty well. I use a thinclient with 4GB emmc as bluetooth/pulse/librespot-connected speaker which is why I didn't want to install linux-firmware on it. Instead, I built a derivation that only copies the relevant bluetooth parts over and called it a day.

> Personally, I'ld love to see zstd support. But as others mentioned xz generally seems to have a better compression ratio than zstd. I hope the suggestions from @aviallon can make a difference.

Agreed.
I hope I'll get to it some time this week :)

Ma27 commented:
Perhaps I misremembered something: I somehow had in mind that high compression levels also affect decompression performance in zstd which is why I stopped after `-15`. However, `-19 --long` may slow down the build-time significantly (though I don't care much about that), but its decompression speed is still way faster (with a naïve timing measurement from `pkgs.time` at least I got 0.02s wallclock time for zstd vs. 0.27s for xz for a 13M blob with 125% (zstd) CPU vs 99% CPU (xz)). I can live with these stats.

The total difference between xz and zstd compressed firmware is now 440M vs 460M, i.e. a size increase of 4.5% (vs 11% before).

Ma27 commented:
So, with `-19 --long` for kernel modules (messed around in the relevant makefile fromt he kernel sourcetree) I got 10% increase of `linuxKernel.kernels.linux_6_1`.
Then I compressed the `.ko` files manually with a dictionary trained on these files before. Then I got it down to a 4% increase (though I'll need to dig a bit into it why it's not even less). That being said, I'll need to understand how the kernel does zstd decompression actually and how we can inject a dictionary into it.

aviallon commented:
It's a pity there is no `Kconfig` option for setting compressor's parameters. Perhaps I could try to tease upstream about that (though I have a feeling it'll get rejected).
In any case, thank you very much @Ma27 for your work 😃

aviallon commented on pkgs/build-support/vm/default.nix:
```diff
@@ -40,10 +40,14 @@ rec {
 
       # Copy BusyBox.
       cp -pd ${pkgs.busybox}/bin/* $out/bin
+      cp -pd ${pkgs.kmod}/bin/* $out/bin
```
Ah, that's right, `busybox` doesn't support zstd modules in its `kmod`.

Ma27 replied:
I think it's not that bad because it's only a build-time dependency in this code-path anyways and NixOS stage1 does kmod as well.

aviallon commented on pkgs/os-specific/linux/kernel/common-config.nix:
```diff
@@ -868,11 +868,13 @@ let
+      ZSWAP                         = option yes;
+      ZSWAP_COMPRESSOR_DEFAULT_ZSTD = whenAtLeast "5.7" yes;
+      ZPOOL                         = yes;
+      ZBUD                          = option yes;
```
```suggestion
      Z3FOLD                          = option yes;
```
Forgot to mention, but I frequently get >3x compression ratios with Zstandard compressed `zram` or `zswap`.
Wich means `ZBUD` (limited to 2x compression ratios) is a bottleneck.

Also, I just discovered [`ZRAM_MULTI_COMP`](https://www.kernel.org/doc/html/latest/admin-guide/blockdev/zram.html#recompression), which is super cool.

K900 replied:
Hm, we should enable that.

alois31 replied:
I don't know about zram, but for zswap at least the default is zsmalloc which should be even better.

aviallon commented on pkgs/build-support/kernel/modules-closure.sh:
```diff
@@ -80,7 +80,7 @@ for module in $(< ~-/closure); do
     modinfo -b $kernel --set-version "$version" -F firmware $module | grep -v '^name:' | while read -r i; do
         echo "firmware for $module: $i"
-        for name in "$i" "$i.xz" ""; do
+        for name in "$i" "$i.xz" "$i.zst" ""; do
             [ -z "$name" ] && echo "WARNING: missing firmware $i for module $module"
             if cp -v --parents --no-preserve=mode lib/firmware/$name "$out" 2>/dev/null; then
```
I know this is unrelated to the PR, but this is such a weird way to write it.
I wonder why it's not someting like:
```bash
if ! cp -v --parents --no-preserve=mode lib/firmware/$i{,.xz,.zst} "$out" 2>/dev/null; then
    echo "WARNING: missing firmware $i for module $module"
fi
```

2xsaiko replied:
That will copy more than one file (though don't know if that can be the case here) and also I think it's really unintuitive here to rely on cp's behavior to keep going if one of the source files doesn't exist (I had to check if it actually does that). Sure, you could write `for name in "$i"{,.xz,.zst}` still, but meh. Above code looks fine to me except for [ instead of [[ (and that's not really something critical).

Ma27 commented:
So, I need a bit of feedback on how to resume here:

##### zram/zswap

ZRAM_MULTI_COMP was enabled in https://github.com/NixOS/nixpkgs/pull/303821#event-12457866752, so we need Z3FOLD here and are done with it.

##### firmware

I think we're good with 4.4% increase. If nobody disagrees, I think this can go to master (along woth the zram/zswap stuff).

##### kernel modules

I think we all agree that the state I have pushed here isn't really acceptable (almost 30% increase of `linuxKernel.kernels.linux_6_1`). There are three options forward:

* Accept 11% increase we get by patching in `-19 --long` (as we do it for firmware blobs) via
  ```
          substituteInPlace scripts/Makefile.modinst \
          --replace-fail '$(ZSTD) -T0' '$(ZSTD) -T0 -19 --long'
  ```
  That'd be the by far quickest solution, but also the dirtiest.
* Let zstd train a dictionary on build that's used to decompress kernel modules. I'm not sure if there's a reasonable way for third-party modules to benefit from that though. However, the zstd implementation of the Linux kernel appears to support dictionaries, it's not used for the module loader however. If we consider this a viable way forward I'd probably try to write a kernel patch for that.
* (Caution: I talk about zstd not in the context of `.ko` files, but in the context of NARs): zstd provides seekable frames (i.e. the ability to decompress portions of a file without decompressing all of it). That could be used under the assumption that between most kernel releases a lot of bytes inside the packages don't change and as a result, Nix could be patched to only fetch ranges of a zstd-compressed NAR that have changed (e.g. by exchanging the hashes of ranges and determining what needs to be re-fetched). That would address the concerns about amount of stuff to download, storage of a binary cache, however it won't account for the disk space of a system.
  (credits to @RaitoBezarius how pointed me towards this idea)

I need a few more opinions from other people because I'm not sure what's the best option to proceed.

Opinions?
cc @lheckemann @k900 @alyssais @RaitoBezarius @fpletz @aviallon @alois31

K900 commented:
I don't think adding delta updates to Nix is realistic at least in the short term. I also think the size increase shouldn't be a huge deal.

Ma27 commented:
@k900 this is correct, I hoped I made it clear that this will take a long time to land (zstd seekable frames for NARs)

fpletz commented:
To shed some light on the other side here: Has anyone checked if we gain any measurable advantages by using zstd? I.e. is compression or decompression significantly quicker? If there are no real advantages, increasing the closure size is IMHO pointless.

RaitoBezarius commented:
I see https://bugs.archlinux.org/task/78892 from the Archlinux folks, some numbers are in there and compare xz. I understood this PR as a way to avoid the xz dependency, due to the recent events.

zstd has the (security?) advantage of being maintained by a large company.

oxalica commented:
> zstd has the (security?) advantage of being maintained by a large company.

I agree it is better maintained than xz, but I have reservation about that general opinion. I'm not surprised by a company claiming they "forced to make data accessible by gov to satisfy a local law endorsement", though it's unlikely to do so in a data formet.

lheckemann commented:
>but its decompression speed is still way faster

The main issue with the really high compression levels isn't speed, but memory usage -- including for decompression -- as far as I understand the manpage.

That aside, I like this!

lheckemann commented on nixos/modules/services/hardware/udev.nix:
```diff
@@ -167,10 +167,17 @@ let
-    pkgs.compressFirmwareXz firmware
-  else
-    id firmware;
+  compressFirmware = firmware:
+    let
+      inherit (config.boot.kernelPackages) kernelAtLeast;
+      compress =
+        if ! (firmware.compressFirmware or true) then
+          id
+        else
+          if kernelAtLeast "5.19" then pkgs.compressFirmwareZstd
+          else if kernelAtLeast "5.3" then pkgs.compressFirmwareXz
+          else id;
+    in compress firmware;
```
nit: no need to wrap this all in an extra lambda
```suggestion
  compressFirmware = let
      inherit (config.boot.kernelPackages) kernelAtLeast;
    in
      if ! (firmware.compressFirmware or true) then
        id
      else
        if kernelAtLeast "5.19" then pkgs.compressFirmwareZstd
        else if kernelAtLeast "5.3" then pkgs.compressFirmwareXz
        else id;
```

Shawn8901 reviewed (approved):
running that since some days, no issues with it, lgtm

lheckemann commented:
I didn't mean to merge the make-modules-30%-bigger change, reverted in #310110. The rest should be fine IMO, it's an improvement overall and leaves less open WIP.

oxalica commented:
Could we reopen this since the change was reverted?

Ma27 commented:
done.
